<launch>

    <!-- Arguments -->
    <arg name="model"             default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
    <arg name="move_forward_only" default="false"/>
    <arg name="scan_topic"        default="scan"/>
    <arg name="remap_goal"        default="false"/>
    <arg name="remap_initialpose" default="false"/>

    <!-- Robot's information -->
    <arg name="multi_robot_name"  default=""/>
    <arg name="initial_pose_x"    default="0.0"/>
    <arg name="initial_pose_y"    default="0.0"/>
    <arg name="initial_pose_a"    default="0.0"/>

    <!-- Turtlebot3 -->
    <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch">
        <arg name="model"            value="$(arg model)" />
        <arg name="multi_robot_name" value="$(arg multi_robot_name)"/>
    </include>

    <!-- amcl -->
    <node pkg="amcl" type="amcl" name="amcl">
        <param name="min_particles"             value="500"/>
        <param name="max_particles"             value="3000"/>
        <param name="kld_err"                   value="0.02"/>
        <param name="update_min_d"              value="0.20"/>
        <param name="update_min_a"              value="0.20"/>
        <param name="resample_interval"         value="1"/>
        <param name="transform_tolerance"       value="0.5"/>
        <param name="recovery_alpha_slow"       value="0.00"/>
        <param name="recovery_alpha_fast"       value="0.00"/>
        <param name="initial_pose_x"            value="$(arg initial_pose_x)"/>
        <param name="initial_pose_y"            value="$(arg initial_pose_y)"/>
        <param name="initial_pose_a"            value="$(arg initial_pose_a)"/>
        <param name="gui_publish_rate"          value="50.0"/>

        <remap from="scan"                      to="$(arg scan_topic)"/>
        <param name="laser_max_range"           value="3.5"/>
        <param name="laser_max_beams"           value="180"/>
        <param name="laser_z_hit"               value="0.5"/>
        <param name="laser_z_short"             value="0.05"/>
        <param name="laser_z_max"               value="0.05"/>
        <param name="laser_z_rand"              value="0.5"/>
        <param name="laser_sigma_hit"           value="0.2"/>
        <param name="laser_lambda_short"        value="0.1"/>
        <param name="laser_likelihood_max_dist" value="2.0"/>
        <param name="laser_model_type"          value="likelihood_field"/>

        <param name="odom_model_type"           value="diff"/>
        <param name="odom_alpha1"               value="0.1"/>
        <param name="odom_alpha2"               value="0.1"/>
        <param name="odom_alpha3"               value="0.1"/>
        <param name="odom_alpha4"               value="0.1"/>
        <!-- <param name="odom_frame_id"             value="odom"/>
        <param name="base_frame_id"             value="base_footprint"/> -->
      
        <param name="odom_frame_id"    value="$(arg multi_robot_name)/odom"/> 
        <param name="base_frame_id"    value="$(arg multi_robot_name)/base_footprint"/>

        <!-- Use the single map server -->
        <remap from="static_map" to="/static_map"/>
        <remap if="$(arg remap_initialpose)" from="initialpose" to="/initialpose"/>
    </node>

    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
        <!-- Set planner -->
        <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS"/>

        <!-- Default configs form the TB3 repos -->
        <rosparam file="$(find turtlebot3_navigation)/param/costmap_common_params_$(arg model).yaml" command="load" ns="global_costmap" />
        <rosparam file="$(find turtlebot3_navigation)/param/costmap_common_params_$(arg model).yaml" command="load" ns="local_costmap" />
        <rosparam file="$(find turtlebot3_navigation)/param/local_costmap_params.yaml" command="load" />
        <rosparam file="$(find turtlebot3_navigation)/param/global_costmap_params.yaml" command="load" />
        <rosparam file="$(find turtlebot3_navigation)/param/move_base_params.yaml" command="load" />
        <rosparam file="$(find turtlebot3_navigation)/param/dwa_local_planner_params_$(arg model).yaml" command="load" />

        <!-- Config with the robot's topics -->
        <param name="global_costmap/scan/sensor_frame"                value="$(arg multi_robot_name)/base_scan"/>
        <param name="global_costmap/obstacle_layer/scan/sensor_frame" value="$(arg multi_robot_name)/base_scan"/>
        <param name="global_costmap/global_frame"                     value="map"/>
        <param name="global_costmap/robot_base_frame"                 value="$(arg multi_robot_name)/base_footprint"/>
        <param name="local_costmap/scan/sensor_frame"                 value="$(arg multi_robot_name)/base_scan"/>
        <param name="local_costmap/obstacle_layer/scan/sensor_frame"  value="$(arg multi_robot_name)/base_scan"/>
        <param name="local_costmap/global_frame"                      value="$(arg multi_robot_name)/odom"/>
        <param name="local_costmap/robot_base_frame"                  value="$(arg multi_robot_name)/base_footprint"/>
      
        <!-- Centralized map server -->
        <remap from="map" to="/map"/>
        <remap if="$(arg remap_goal)" from="move_base_simple/goal" to="/move_base_simple/goal"/>
        <remap if="$(arg remap_initialpose)" from="initialpose" to="/$(arg multi_robot_name)/initialpose"/>
    </node>
</launch>
